frontend/templates/product/hooks/useSelectedVariant.ts

```diff
@@ -1,70 +1,68 @@
 import {
-   invariant,
-   getVariantIdFromVariantURI,
    getProductVariantURI,
+   getVariantIdFromVariantURI,
 } from '@ifixit/helpers';
 import type { Product, ProductVariant } from '@pages/api/nextjs/cache/product';
-import { useRouter } from 'next/router';
-import { useCallback } from 'react';
+import { useRouter, useSearchParams } from 'next/navigation';
 
-type SetVariantIdFn = (variantId: string) => void;
+type SelectVariantFn = (variantId: string) => void;
 
 export function useSelectedVariant(
    product: Product
-): [ProductVariant, SetVariantIdFn] {
+): [ProductVariant, SelectVariantFn] {
    const router = useRouter();
+   const searchParams = useSearchParams();
 
-   const defaultVariantId = useDefaultVariantId(product);
-   const searchParamVariantId = useSearchParamVariantId();
+   const selectVariant = (variantId: string) => {
+      router.replace(variantUrlFor(variantId), { scroll: false });
+   };
 
-   const currentVariantId = searchParamVariantId ?? defaultVariantId;
+   return [selectedVariant(), selectVariant];
 
-   let variant = product.variants.find((v) => v.id === currentVariantId);
-   if (variant == null) {
-      variant = product.variants.find((v) => v.id === defaultVariantId);
+   function selectedVariant() {
+      return searchParamVariant() ?? defaultVariant();
    }
 
-   invariant(
-      variant,
-      `Something went wrong, variant with id "${currentVariantId}" not found`
-   );
+   function searchParamVariant() {
+      if (variantGIDParam() == null) return null;
+      return product.variants.find((v) => v.id === variantGIDParam()) ?? null;
+   }
 
-   const setVariantId = useCallback<SetVariantIdFn>(
-      (variantId) => {
-         const { variant, ...newQuery } = router.query;
-         if (variantId !== defaultVariantId) {
-            newQuery.variant = getVariantIdFromVariantURI(variantId);
-         }
-         router.replace(
-            {
-               query: newQuery,
-            },
-            undefined,
-            { shallow: true }
-         );
-      },
-      [defaultVariantId, router]
-   );
+   function defaultVariant() {
+      return (
+         product.variants.find(
+            (variant) =>
+               variant.quantityAvailable && variant.quantityAvailable > 0
+         ) ?? product.variants[0]
+      );
+   }
```

Comment:
We could reuse `defaultVariantIdFor` so as not to duplicate this logic. It isn't terribly complex, but it's nice to have a single source of truth for what determines the default variant.

Reply:
The `app/` folder enables better code colocation, so I wanted to move all this stuff into the product folder after we ship. This is not the final place for this code, just the minimal set of changes in order to ship. 
If you prefer to refactor this to use `defaultVariantIdFor` within the context of this PR let me know 👍

Reply:
Since there's no CR yet I just did the change 🚢

Reply:
> This is not the final place for this code, just the minimal set of changes in order to ship.

I see, sorry for nitpicking! I forget you have been staging these PR's for easier review.

Reply:
There was always the risk to forget this, so you were right calling this out 👍